src/functional/list.c

```diff
@@ -0,0 +1,53 @@
+#include "list.h"
+#include "gc.h"
+
+list_t* cons(void* head, void* tail) {
+	list_t* list = gc_malloc(sizeof *list);
+
+	return list ? (*list = (list_t) { head, tail }, list) : NULL;
+}
+
+void* head(list_t* list) {
+	return list->head; 
+}
+
+void* tail(list_t* list) {
+	return list->tail; 
+}
+
+bool is_empty(list_t* list) {
+	return list == NULL;
+}
+
+list_t* concat(list_t* left, list_t* right) {
+	return is_empty(left)
+		? right
+		: cons(head(left), concat(tail(left), right));
+}
+
+void* nth(size_t index, list_t* list) {
+	return index == 0 ? head(list) : nth(index - 1, tail(list));
```

Comment:
if we support only prominent compilers `gcc/clang` then I'll add `-foptimize-sibling-calls` compiler flag to `configure.ac` it will replace recursion with iterations. Otherwise I could do it for compiler - change recursion to loops.